### PR TITLE
[recipe, megatron, fsdp] feat: Support fully async on NPU

### PR DESCRIPTION
**Pull request by Shangwei-Li**
### What does this PR do?

Due to ray collective not supported on NPU yet, this PR uses vllm stateless group to replace ray collective group to support fully async training and checkpoint engine.

### Checklist Before Starting

- [x] Search for similar PRs. Paste at least one query link here: https://github.com/volcengine/verl/pulls?q=fully+async+npu
- [x] Format the PR title as `[{modules}] {type}: {description}` (This will be checked by the CI)
  - `{modules}` include `fsdp`, `megatron`, `sglang`, `vllm`, `rollout`, `trainer`, `ci`, `training_utils`, `recipe`, `hardware`, `deployment`, `ray`, `worker`, `single_controller`, `misc`, `perf`, `model`, `algo`, `env`, `tool`, `ckpt`, `doc`, `data`, `cfg`, `reward`
  - If this PR involves multiple modules, separate them with `,` like `[megatron, fsdp, doc]`
  - `{type}` is in `feat`, `fix`, `refactor`, `chore`, `test`
  - If this PR breaks any API (CLI arguments, config, function signature, etc.), add `[BREAKING]` to the beginning of the title.
  - Example: `[BREAKING][fsdp, megatron] feat: dynamic batching`

### Test

> For changes that can not be tested by CI (e.g., algorithm implementation, new model support), validate by experiment(s) and show results like training curve plots, evaluation results, etc.

We have looked into the performance and throughput of fully async with Qwen3-30B-A3B. 
With FSDP backend, we have achieved even higher reward than a previous collocated training. (The pink line represent fully async while the others representing collocated)
image
And the validation results matches gpu results(max error<2%. The red line represents NPU, orange line represents GPU.).
image

As for throughput, we have achieved more than 4x throughput gaining comparing 64-rank fully async training to 128-rank collocated training, which means 64-rank fully async training is 2x faster than collocated  128-rank.
image

Also verified with Qwen3-0.6B in cases below:

1. Megatron backend.
2. Checkpoint engine disabled.

### API and Usage Example

> Demonstrate how the API changes if any, and provide usage example(s) if possible.

```python
# Add code snippet or script demonstrating how to use this
```

### Design & Code Changes

> Demonstrate the high-level design if this PR is complex, and list the specific changes.

### Checklist Before Submitting

> [!IMPORTANT]
> Please check all the following items before requesting a review, otherwise the reviewer might deprioritize this PR for review.

- [x] Read the [Contribute Guide](https://github.com/volcengine/verl/blob/main/CONTRIBUTING.md).
- [x] Apply [pre-commit checks](https://github.com/volcengine/verl/blob/main/CONTRIBUTING.md#code-linting-and-formatting): `pre-commit install && pre-commit run --all-files --show-diff-on-failure --color=always`
- [x] Add / Update [the documentation](https://github.com/volcengine/verl/tree/main/docs).
- [ ] Add unit or end-to-end test(s) to [the CI workflow](https://github.com/volcengine/verl/tree/main/.github/workflows) to cover all the code. If not feasible, explain why: ...
- [x] Once your PR is ready for CI, send a message in [the `ci-request` channel](https://verl-project.slack.com/archives/C091TCESWB1) in [the `verl` Slack workspace](https://join.slack.com/t/verl-project/shared_invite/zt-3855yhg8g-CTkqXu~hKojPCmo7k_yXTQ). (If not accessible, please try [the Feishu group (飞书群)](https://applink.larkoffice.com/client/chat/chatter/add_by_link?link_token=772jd4f1-cd91-441e-a820-498c6614126a).)

### CODE REVIEW

File: recipe/fully_async_policy/checkpoint_engine.py [hunk 1]
```diff
@@ -169,7 +168,11 @@ def get_ip() -> str:
 def npu_generate_uuid() -> str:
     """Generate uuid for each npu device"""
     str_pid = str(os.getpid())
-    npu_num = 8
+    # In A2 server, one x86 machine could have more than 8 cards.
+    if "910C" not in torch.npu.get_device_name():
```
Comment (tardis-key) [inline recipe/fully_async_policy/checkpoint_engine.py]: try to avoid specifying the chip model in the code

File: recipe/fully_async_policy/checkpoint_engine.py [hunk 2]
```diff
@@ -182,6 +185,8 @@ def npu_generate_uuid() -> str:
                 search_after_pid = str_result[str_result.find(str_pid) + len(str_pid) :]
                 match_chip_id = re.search(r"Chip ID[^\d]*(\d+)", search_after_pid)
                 chip_id = int(match_chip_id.group(1))
+                # NOTE: get_ip could possibly fail due to hostname not bound to an IP address.
```
Comment (tardis-key) [inline recipe/fully_async_policy/checkpoint_engine.py]: file an issue instead of writing a note in codes

File: recipe/fully_async_policy/checkpoint_engine.py [hunk 3]
```diff
@@ -169,7 +168,11 @@ def get_ip() -> str:
 def npu_generate_uuid() -> str:
     """Generate uuid for each npu device"""
     str_pid = str(os.getpid())
-    npu_num = 8
+    # In A2 server, one x86 machine could have more than 8 cards.
+    if "910C" not in torch.npu.get_device_name():
+        npu_num = torch.npu.device_count()
```
Comment (FightingZhen) [inline recipe/fully_async_policy/checkpoint_engine.py]: Just use `torch.npu.device_count()` to get real device num, considering `910C` is not required.

File: recipe/fully_async_policy/checkpoint_engine.py (hunk 2 above)
Comment (FightingZhen) [inline recipe/fully_async_policy/checkpoint_engine.py]: If `get_ip()` is not rubost enough but have no extra solutions, it is not advised to add comments into code, create a issue instead.

File: recipe/fully_async_policy/checkpoint_engine.py [hunk 4]
```diff
@@ -479,6 +490,7 @@ def update_weights_from_ipc_(socket_path):
 
         gidx = 0
         local_buckets = self.global_buckets.get(self.current_rank, [])
+        dummy_tensor = torch.tensor([1.0], device=get_torch_device().current_device())
```
Comment (FightingZhen) [inline recipe/fully_async_policy/checkpoint_engine.py]: Use `verl.utils.device.get_device_id()` instead.

File: recipe/fully_async_policy/distributed_util.py [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+from datetime import timedelta
+
+from torch.distributed import TCPStore
+
+from verl.utils.device import is_npu_available
+
+
+def is_ipv6(address: str) -> bool:
+    """
+    判断字符串是否为合法的IPv6地址
```
Comment (FightingZhen) [inline recipe/fully_async_policy/distributed_util.py]: Use English comments

File: recipe/fully_async_policy/distributed_util.py [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+    # the following can be used:
+    # from sglang.srt.distributed.device_communicators.pynccl import PyNcclCommunicator
+    # from sglang.srt.distributed.utils import statelessprocessgroup
+    if is_npu_available:
+        from vllm_ascend.distributed.device_communicators.pyhccl import PyHcclCommunicator as PyNcclCommunicator
+    else:
+        from vllm.distributed.device_communicators.pynccl import PyNcclCommunicator
+    from vllm.distributed.utils import StatelessProcessGroup
+
+    def my_create(
```
Comment (FightingZhen) [inline recipe/fully_async_policy/distributed_util.py]: `my_create()` is not a good func name, maybe rename it to `create_process_group`？

File: recipe/fully_async_policy/fsdp_workers.py [hunk 7]
```diff
@@ -155,7 +161,9 @@ def sync_rollout_weights_by_checkpoint(self, sync_group_name="actor_rollout"):
         register_duration = register_end_time - cache_end_time
         self.cpu_named_params = {}
 
-        collective.barrier(group_name=sync_group_name)
+        # collective.barrier(group_name=sync_group_name)
```
Comment (FightingZhen) [inline recipe/fully_async_policy/fsdp_workers.py]: Remove unused code.

File: recipe/fully_async_policy/fsdp_workers.py [hunk 8]
```diff
@@ -155,7 +161,9 @@ def sync_rollout_weights_by_checkpoint(self, sync_group_name="actor_rollout"):
...
+        dummy_tensor = torch.tensor([1.0], device=get_torch_device().current_device())
```
Comment (FightingZhen) [inline recipe/fully_async_policy/fsdp_workers.py]: Use `verl.utils.device.get_device_id()` instead, same as follows.

File: recipe/fully_async_policy/distributed_util.py [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+
+    def create_process_group(
+        host: str,
+        port: int,
+        rank: int,
+        world_size: int,
+        data_expiration_seconds: int = 3600,
+        store_timeout: int = 300,
+    ) -> "StatelessProcessGroup":
+        """This is copied from vllm project to support ipv6 stateless communication groups."""
```
Comment (FightingZhen) [inline recipe/fully_async_policy/distributed_util.py]: Suggest adding comments about vLLM original code file address

File: recipe/fully_async_policy/megatron_worker.py [hunk 10]
```diff
@@ -198,6 +206,20 @@ def sync_rollout_weights_by_checkpoint(self, sync_group_name="actor_rollout"):
                 f" offload model to cpu cost {offload_duration} seconds"
             )
 
+    @register(dispatch_mode=Dispatch.ONE_TO_ALL, blocking=False)
```
Comment (FightingZhen) [inline recipe/fully_async_policy/megatron_worker.py:210]: This function looks same in fsdp/megatron worker, can we move it to a public place and use it in fsdp/megatron backend?

Comment (Shangwei-Li) [reply recipe/fully_async_policy/megatron_worker.py:210]: This could be easily achieved after this recipe migrates to engine workers. If we move it to Worker class now it would result in another class maintained in recipe.

**Comment by Shangwei-Li** [comment]
Close this pr as https://github.com/volcengine/verl/pull/4885/files is open.
